test/helpers/domain/domain.ts

```diff
@@ -0,0 +1,287 @@
+  tokenURI : string;
+
+  constructor ({
+    zns,
+    domainConfig,
+  } : {
+    zns : IZNSContracts;
+    domainConfig : IFullDomainConfig;
+  }) {
+    this.zns = zns;
+
+    this.owner = domainConfig.owner;
+    this.parentHash = domainConfig.parentHash || hre.ethers.ZeroHash;
+    this.isRoot = this.parentHash === hre.ethers.ZeroHash;
+    this.label = domainConfig.label;
+    this.tokenOwner = domainConfig.tokenOwner || hre.ethers.ZeroAddress;
+    this.distrConfig = domainConfig.distrConfig || distrConfigEmpty;
+
+    switch (this.distrConfig.pricerContract) {
+    case zns.curvePricer.target:
+      this.priceConfig = curvePriceConfigEmpty;
+      break;
+    case zns.fixedPricer.target:
+      this.priceConfig = fixedPriceConfigEmpty;
+      break;
+    default:
+      this.priceConfig = {} as ICurvePriceConfig | IFixedPriceConfig;
+    }
+
+    this.paymentConfig = domainConfig.paymentConfig || paymentConfigEmpty;
+    this.domainAddress = domainConfig.domainAddress || this.owner.address;
+    this.tokenURI = domainConfig.tokenURI || "https://example.com/token-uri";
+
+    this.hash = "";
+  }
+
+  get tokenId () : bigint {
+    return BigInt(this.hash);
+  }
+
+  async ownerOfHash () : Promise<string> {
+    return this.zns.registry.getDomainOwner(this.hash);
+  }
+
+  async ownerOfToken () : Promise<string> {
+    return this.zns.domainToken.ownerOf(this.tokenId);
+  }
+
+  async getDomainHashFromEvent (domainOwner ?: SignerWithAddress) : Promise<string> {
+    const latestBlock = await time.latestBlock();
+    const filter = this.zns.rootRegistrar.filters.DomainRegistered(
+      undefined,
+      undefined,
+      undefined,
+      undefined,
+      undefined,
+      domainOwner ? domainOwner : this.owner.address,
+      this.tokenOwner === hre.ethers.ZeroAddress ? undefined : this.tokenOwner,
+      undefined,
+    );
+
+    const events = await this.zns.rootRegistrar.queryFilter(filter, latestBlock - 2, latestBlock);
+    const { args: { domainHash } } = events[events.length - 1];
+
+    return domainHash;
+  }
+
+  async mintAndApproveForDomain (user ?: SignerWithAddress) : Promise<ContractTransactionResponse> {
+    return (fundApprove({
+      zns: this.zns,
+      parentHash: this.parentHash,
+      user: user ? user : this.owner,
+      domainLabel: this.label,
+    })) as Promise<ContractTransactionResponse>;
+  }
+
+  async register (executor ?: SignerWithAddress) : Promise<ContractTransactionResponse> {
+    const {
+      zns,
+      owner,
+      label,
+      parentHash,
+      distrConfig,
+      paymentConfig,
+      tokenURI,
+      tokenOwner,
+      domainAddress,
+    } = this;
+
+    if (!distrConfig.priceConfig) {
+      distrConfig.priceConfig = encodePriceConfig(this.priceConfig);
+    }
+
+    let txPromise : ContractTransactionResponse;
+
+    // mint and approve strict amount of tokens for domain registration
+    await this.mintAndApproveForDomain(executor);
+
+    if (this.isRoot) {
```

Comment:
Why bother doing all the deconstruction of `this` above on L110-L120 when you just access using `this.[prop_name]` anyways?